rfcs/0001-meta-schedule-autotensorir.md

```diff
@@ -0,0 +1,444 @@
+l14, l15, l16, l17 = sch.split(loop=l1, factors=[v4, v5, v6, v7])
+l18, l19, l20, l21 = sch.split(loop=l2, factors=[v8, v9, v10, v11])
+l22, l23 = sch.split(loop=l3, factors=[v12, v13])
+sch.reorder(l14, l18, l15, l19, l22, l16, l20, l23, l17, l21)
+```
+
+### 4.2. Exploring the Design Space
+
+Meta Schedule provides several built-in exploration strategies to exhaustively or efficiently search
+for efficient schedules.
+
+**Random search by replaying schedule functions.** With a user-provided schedule function
+as a black-box design space generator, our system could repetitively invoke such an opaque function
+without doing any extra analysis. The function could be written in C++ or Python, or any language
+that implements packed function FFI. If sampling instructions are present in the function, then each
+invocation results in a different IRModule after being scheduled because the random decisions are
+possibly changed across different runs. Effectively, it is equivalent to
+random exploration without trace, allowing the flexibility for users to define arbitrary functions
+that trace may not well support (e.g. control flow divergence based on the value of intermediate
+random variables), but it forbids future opportunity of any trace-based analysis.
+
+**Random search by replaying traces.** Traces are obtained from a design space generator, and
+replayed with a builtin interpreter in our system. If sampling instructions are present on the
+traces, then their random decisions are mutated during each replay, i.e. jumps to a new point in the
+design space. Therefore, repetitive replay of those traces are equivalent to exploration of the
+design space. Our system could potentially benefit from trace-based analysis, including rejecting
+obviously invalid schedules (e.g. using too much CUDA resources), doing dead-code elimination to
+simplify a trace, extracting trace-based features used in the cost model, etc.
+
+**Cost-model-guided evolutionary search**. A more efficient exploration strategy. We define two sets
+of rules:
+
+* Mutator: defines how to jump to a point’s "neighbor" in the design space
+* Postprocessor: after the trace is executed, there are some extra rules we want to execute, for
+  example:
+  * Check CUDA resource limits: There is a hard requirement in CUDA that the maximum number of
+    threads should not exceed 1024, but it is a random variable that cannot be determined before
+    actually executing the trace. In this case, we write a postprocessor that errors out when the
+    condition is not satisfied.
+  * Fuse outer loops until the extent of the fused loops is large enough: The number of outer loops
+    to be fused together depends on their extents, which are random variables. In this case, we
+    annotate the maximum extent allowed on the block, and do actual fusion in a postprocessor.
+
+Our evolutionary search algorithm uses mutators to find possible schedules in the design space, then
```

Comment:
i think this process might be best explained with a numbered list. also, can you provide more context including what the process starts and ends with?

The evolutionary search process is as follows:
1. Begin with a naive TIR schedule
2. Apply Mutator to find possible schedule transformations
3. Pick one (how?)
4. Apply post-processor to the choice to validate/tweak it
5. Measure or predict performance with cost model (explain when you would do one or the other)
6. Repeat steps 2-6 until <exit condition>

Reply:
It is a very complicated process, but almost identical to Ansor. I dropped a citation to the Section 5.1 in Ansor's paper for more details